src/alire/alire-environment-formatting.adb

```diff
@@ -0,0 +1,127 @@
+with Ada.Strings.Unbounded; use Ada.Strings.Unbounded;
+
+with Alire.OS_Lib;
+with Alire.Directories;
+with Alire.Platform;
+with Alire.Paths;
+
+package body Alire.Environment.Formatting is
+
+   ----------------
+   -- Find_Start --
+   ----------------
+
+   function Find_Start (Str  : Unbounded_String;
+                        From : Positive)
+                        return Natural
+   is
+      Loc : Natural := From;
+   begin
+
+      loop
+         Loc := Index (Str, "${", Loc);
+
+         exit when Loc = 0;
+
+         --  Check for escape character before the pattern
+         if Loc = 1 or else Element (Str, Loc - 1) /= '\' then
+            return Loc;
+         else
+
+            --  Pattern found but with an escape character
+            Loc := Loc + 2;
+         end if;
+      end loop;
+
+      return Loc;
+   end Find_Start;
+
+   --------------
+   -- Find_End --
+   --------------
+
+   function Find_End (Str  : Unbounded_String;
+                      From : Positive)
+                      return Natural
+   is
+   begin
+      --  There is not possible escape character between the start and
+      --  the end of the formating pattern.
+      return Index (Str, "}", From);
+   end Find_End;
+
+   ------------
+   -- Format --
+   ------------
+
+   function Format (Rel             : Releases.Release;
+                    Value           : String;
+                    Is_Root_Release : Boolean)
+                    return String
+   is
+      -------------
+      -- Replace --
+      -------------
+
+      procedure Replace (Str : in out Unbounded_String;
+                         From, To : Positive)
+      is
+         use Alire.OS_Lib;
+
+         Id : constant String := Slice (Str, From + 2, To - 1);
+
+         Working_Folder : constant Alire.Absolute_Path :=
+           Alire.Directories.Current;
+      begin
+
+         if Id = "DISTRIB_ROOT" then
+            Replace_Slice (Str, From, To, Platform.Distribution_Root);
+
+         elsif Id = "CRATE_ROOT" then
+            Replace_Slice
+              (Str, From, To,
+               Working_Folder /
```

Comment:
I guess the way I compute the root directory of the crate is not compatible with softlink pins.

If possible it would be best to have a function to get that info on any release.

Reply:
Right. Let me take a stab at this.